pkg/operation/botanist/vpnseedserver.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+
+	"github.com/gardener/gardener/charts"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpnseedserver"
+	"github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+
+	"k8s.io/utils/pointer"
+)
+
+const diffieHellmanKey = "LS0tLS1CRUdJTiBESCBQQVJBTUVURVJTLS0tLS0KTUlJQkNBS0NBUUVBN2NCWHhHOWFuNktSei9zQjV1aVNPVGY3RWcrdVdWa2hYTzRwZUtEVEFSek1ZYThiN1dSOApCL0F3K0F5VVh0QjN0WHRyemVDNU0zSUhudWhGd01vM0s0b1NPa0ZKeGF0TGxZS2VZMTVyK0t0NXZuT09UM0JXCmVONU9uV2xSNVdpN0daQldiYVFnWFZSNzlONHlzdDQzc1ZoSnVzNkJ5MGxONk9sYzl4RC95czlHSC95a0pWSWgKWi9OTHJ4QUM1bHhqd0NxSk1kOGhycnlDaHVEbHo1OTd2ZzZnWUZ1UlY2MFUvWVU0REs3MUY0SDdtSTA3YUdKOQpsK1NLOFRia0tXRjVJVEk3a1lXYmM0em10ZlhTWGFHak1oTTlvbVFVYVRIOWNzQjk2aHpGSmRlWjRYanh5YlJmClZjM3Q3WFA1cTdhZmVhS21NM0ZoU1hkZUhLQ1RxUXpRdXdJQkFnPT0KLS0tLS1FTkQgREggUEFSQU1FVEVSUy0tLS0tCg=="
```

Comment:
Why is this hard-coded?

Reply:
It is documented in https://github.com/gardener/gardener/blob/master/docs/usage/configuration.md that there is a default diffie hellman key for the open vpn solution. This default key was stored in the values.yaml in the past. As we were told to not use charts anymore, we have to store the default key somewhere else.
Do you have a better proposal where to store it?

Reply:
There should not be a default key. It is generally ok to use a key for various shoot clusters (or even all the shoots for a particular seed) but there should be no static default.

Reply:
Back then, we refrained from generating such key for each shoot because it was costly, if I remember correctly. For productive installations, as referenced by @ScheererJ, it's anyways recommended to configure an own key. In this case, the default is not used. However, it simplifies the overall deployment complexity if we do not make generating a DH key a new requirement. Can we simply keep it as is because this is also how it was before/the past 3-4 years?

Reply:
Thanks @rfranzke, I do remember that conversation from almost precisely 3 years ago. At that time we got the following recommendation:

> Some words of warning!
> openssl dhparam -text 2048 dose nothing beyond searching for ONE good / suitable DH prime number.
> You have a small Generator value g == 2 per default and need the check for safe primes.
>
> The speed is bound to your random-pool (/dev/random) and not to CPU or Memory.
> Such DH parameters are designed to be reused.
 
> Do NOT use openssl dhparam -dsaparam for reuse purpose!!! This will generate a random large Generator value g but omit the check for safe primes.
> The speed difference is that you do not have to generate so many primes because you don’t throw away all the unsafe ones.
> 
> Therefore my recommendation matches the last sentence.
> Generate the DH prime on a PC with good random generator and simply reuse it in your landscape. This is even simpler, and it is safe.

So I need to correct myself: it is ok to keep that value. Sorry for the confusion.